`UsedForField` widget on sighting

sandreae opened the pull request:
## Implemented
- `UsedFor` model and associated queries 
- `UsedForField` widget for adding existing "used for" tags to a sighting, or creating new ones and directly adding them.
- makes use of `InfiniteScrollList` and `InfiniteDedupTagsList` which both handle making paginated requests, showing the results in a list, and refreshing the query on scroll to bottom. There is a lot of code duplication in these components (because of the (ugly) dedup logic required when querying all uses, see below for details) but I couldn't see a better way so far.

## Possible Improvements
- design can definitely be improved
- maybe add a button for submitting new uses, rather than clicking the top level submit button (which then closes edit mode)
- think of a way to not require the dedup logic when querying all uses. One way is as described in the [comments below](https://github.com/p2panda/meli/pull/48#discussion_r1365748426), to change our schema so that "uses" are related to from a sighting using a join document, rather than themselves containing the relation to the sighting. Then we could query for all uses documents when we want to display all tags, and query for joins when we want to show uses for one sighting. This is all optional though, and depends on how much we hate the de-duplication and/or if this same challenge occurs elsewhere



[Screen_recording_20231019_161022.webm](https://github.com/p2panda/meli/assets/24291662/8768f356-ea26-4216-9dce-a23d7e7e7705)

(this vid is slightly out of date as the autocomplete widget is no longer used for text input, otherwise it's the same)

adzialocha commented on packages/app/lib/ui/screens/sighting.dart:
```diff
@@ -118,6 +121,23 @@ class _SightingProfileState extends State<SightingProfile> {
     setState(() {});
   }
 
+  Future<DocumentViewId?> _updateUsedFor(AutocompleteItem? item) async {
```
Return value can be void

adzialocha commented on packages/app/lib/ui/screens/sighting.dart:
```diff
@@ -139,6 +159,11 @@ class _SightingProfileState extends State<SightingProfile> {
           onUpdate: _updateSpecies,
         ),
+        UsedForField(
+          null,
```
Null!

sandreae commented on packages/app/lib/ui/widgets/pagination_used_for_tags_list.dart:
```diff
@@ -0,0 +1,148 @@
+          }
+        };
+
+        // Deduplicate the returned collection.
+        var seen = Set<String>();
+        List<UsedFor> uniqueUses = data.documents
+            .where((usedFor) => seen.add(usedFor.usedFor))
+            .toList();
```
In the next line I account for the situation you describe by re-requesting items until we get a certain amount of unique uses. (still a hack on a hack)...

https://github.com/p2panda/meli/blob/ac0ab64edc5107e46c58f1f95fe00d43bbaeb100/packages/app/lib/ui/widgets/pagination_used_for_tags_list.dart#L132-L137

But yeh, totally agree that if we re-think the data type design a bit we could really make this nicer. 

I think the a really nice way of modelling this would be to move the sighting relation field into it's own "join" document: 

```
BeeAttributeUsedFor {
  description: str
}

BeeAttributeAssociation {
  attribute: relation(bee_attribute_used_for)
  relation: relation(bee_sightning)
}
```

This way we get;
- better re-use of UsedFor attributes
- nice aggregates over UsedFor attributes (`totalCount`) 
- familiar pattern in the app for querying, displaying, auto-completing attributes (no dedup needed)
- if you delete an "association" it doesn't actually delete the attribute

sandreae replied:
Any change like this we make we would want to do for the other attributes (location).

sandreae replied:
We _could_ combine the attribute association for all attributes into one like so:

```
BeeAttributeAssociation {
  attribute: relation(bee_attribute_used_for)
  sighting: relation(bee_sightning)
}

```

sandreae replied:
We end up with even more documents and schema :rofl: but maybe that's just unavoidable with any kind of meaningful application.

sandreae replied:
Alternatively we do it all like `local_name` (relation list with empty list representing no value) and add attribute fields for every possible attribute to bee sighting. 

Here we also get all the nice repeatable behaviour around the attributes but with fewer schema (just a big sighting).

adzialocha replied:
> Any change like this we make we would want to do for the other attributes (location).

Wondering what data we're interested in re-using and which not? Or where do we need extra value from knowing that data appeared across sightings?

With "location" I think its fine as we've designed it, as it is really only an attribute of that specific sighting.

sandreae replied:
Ah yeh, that's true, locations aren't reused in the same way, that's cool :+1:

sandreae replied:
What do you think of adding the join document then? Seems quite worth it to me.

adzialocha replied:
> What do you think of adding the join document then? Seems quite worth it to me.

It's cool! I wouldn't introduce it just for "Used for" maybe? Or is there a reason this data behaves differently than "local name"? Maybe use the same pattern for both?

sandreae replied:
I think the difference between "local_name" and "used_for" is that with the former we are only wanting one value per sighting, whereas with the later we want a list.

sandreae replied:
Of course, "local_name" is actually a list 🙃 but in terms of how we allow users to interact with the data, this is the difference.

sandreae replied:
Is this a fundamental difference wrt this conversation about the data design?... I'm not actually sure...

sandreae replied:
We did talk about allowing "local_name" to be a list at one point, as different people might know the same bee by different local names.

adzialocha reviewed (changes requested):
Thank you for this PR, this one was really tricky!

I've just tested this on my app and it _almost_ works. This is what I've observed:

1. Whenever I look at it for the first time it is unclear what each of these fields stand for and what I can do with them. It's also unclear where the input field is to add something.

![photo_2024-05-07_18-03-11](https://github.com/p2panda/meli/assets/1822473/c92e734f-6ca3-4bf9-9036-31baf4de3d50)

My suggestion would be to a) add titles to each of these three areas and b) introduce a "Create" button, similar to what we have with the "Notes" widget?

2. Not clear that one can scroll down the field, can we add a scrollbar here?

![photo_2024-05-07_18-03-09](https://github.com/p2panda/meli/assets/1822473/e75f857b-91ae-4cd6-9ed5-120a156ebf2f)

3. Error in GraphQL query?

Not sure what it means but I receive this here:

> E/flutter ( 9891): [ERROR:flutter/runtime/dart_vm_initializer.cc(41)] Unhandled Exception: Exception: Query is not refetch safe
E/flutter ( 9891): #0      ObservableQuery.refetch (package:graphql/src/core/observable_query.dart:142:5)
E/flutter ( 9891): #1      _UsedForFieldState._createUse (package:app/ui/widgets/used_for_field.dart:63:26)
E/flutter ( 9891): <asynchronous suspension>
E/flutter ( 9891): #2      _UsedForFieldState._onTagClick (package:app/ui/widgets/used_for_field.dart:138:5)
E/flutter ( 9891): <asynchronous suspension>
E/flutter ( 9891): 

Otherwise it looks fine to me! Could definitely need a little bit of design touches (less padding etc.) but don't think it's much left to make it work already.

> One way is as described in the https://github.com/p2panda/meli/pull/48#discussion_r1365748426, to change our schema so that "uses" are related to from a sighting using a join document, rather than themselves containing the relation to the sighting

I like that! Should we go for it? It could really help with the aggregations as well ..

sandreae commented:
Thanks @adzialocha, nice suggestions. I made most of those adjustments and refactored everything into smaller widgets, all much nicer now. 

> Whenever I look at it for the first time it is unclear what each of these fields stand for and what I can do with them. It's also unclear where the input field is to add something.
>
> My suggestion would be to a) add titles to each of these three areas and b) introduce a "Create" button, similar to what we have with the "Notes" widget?

Yeh, I did all this, and improved the messages one sees when there are no entries in the lists.

> Not clear that one can scroll down the field, can we add a scrollbar here?

Yip, added.

> Error in GraphQL query?

I have actually seen this message previously, but then it suddenly stopped... Haven't seen it at all since doing a refactor of the widgets, so maybe that helped somehow... Do you still see it now?

> Could definitely need a little bit of design touches (less padding etc.) 

I spent a good 20mins trying to reduce the padding but can't for the life of me figure out where it's coming from :sob: 

> > One way is as described in the https://github.com/p2panda/meli/pull/48#discussion_r1365748426, to change our schema so that "uses" are related to from a sighting using a join document, rather than themselves containing the relation to the sighting
> 
> I like that! Should we go for it? It could really help with the aggregations as well ..
> 

I will look more into this, but want to have a better idea what is involved in aggregating over uses per species as well. I think if we make the suggested change here, then we will still need to dedup in that case anyway..... Let me explore more in following steps.

adzialocha reviewed (approved):
Great, thank you for the adjustments @sandreae! Much better now, especially hiding the UI when being in "read" mode is nice. This is good to merge from my side.
